### PR TITLE
Added schema function

### PR DESCRIPTION
**Pull request by a2wd**
* Added schema functions to each of `ICommonContext`, `Table`, `Operation` and `Assertion`.
* Added corresponding tests
* Refactored the resolve method to de-dupe common logic in normalizing the `Schema`/`Database` and `Table` names.
* Removed redundant calls to the `toResolvable` method

### CODE REVIEW

File: tests/core/core.spec.ts [hunk 1]
```diff
@@ -887,6 +887,37 @@ suite("@dataform/core", () => {
         });
       }
     );
+
+    test("schema without suffix", () => {
+      const session = new Session(path.dirname(__filename), TestConfigs.redshift);
+      session.publish("test", { type: "table" })
+        .query(ctx => `select "${ctx.schema()}"`);
```
Comment (Ekrekr) [inline tests/core/core.spec.ts:894]: Nit: can just be `ctx => ctx.schema()`

**Review by Ekrekr — COMMENTED** [review]
Nice, looks great to me! But wait for ben to check.

Probably worth doing a patch version bump https://github.com/dataform-co/dataform/blob/main/version.bzl

Nit: do the refactoring part in a separate CL next time

File: core/session.ts [hunk 2]
```diff
@@ -444,6 +431,21 @@ export class Session {
...
+    return this.adapter().normalizeIdentifier(
+      `${database}${this.getDatabaseSuffixWithUnderscore()}`);
+  }
+
+  public normalizeSchema(schema: string): string {
+    return this.adapter().normalizeIdentifier(
+      `${schema}${this.getSchemaSuffixWithUnderscore()}`);
+  }
+
+  public normalizeTable(table: string): string {
```
Comment (BenBirt) [inline core/session.ts:444]: nit: i think "normalize" isn't quite the right word to use here. elsewhere in the code, "normalizing" an identifier means: to translate (usually involves escaping, casing, whatever) this string to something which can be used as an identifier in the target warehouse.

here we *are* normalizing, sure, but we're doing more than that - we are suffixing/prefixing as necessary.

i would suggest "finalize", i.e. `finalizeDatabase`

Comment (BenBirt) [inline core/session.ts:444]: this should be `normalizeName`. i realise that "name" as a term really sucks, but it is what the entire framework (and other code which uses the framework) uses to indicate the last part of a target, which could be a table, view, or any other materializable object - or might not even be materializable, in the case of `operations`!

File: tests/core/core.spec.ts [hunk 3]
```diff
@@ -887,6 +887,37 @@ suite("@dataform/core", () => {
         });
       }
     );
+
+    test("schema without suffix", () => {
```
Comment (BenBirt) [inline tests/core/core.spec.ts:891]: generally for suffix/suffix-less tests we do some table-driven stuff, i.e. something like

```
for ({config, expectedTarget of [{config: TestConfigs.redshift, expectedTarget: "schema.test"}, ...]) {
  test("schema with config " + JSON.stringify(config), () => {
    // test
  }
}
```

**Review by BenBirt — COMMENTED** [review]
Please bump the minor version in `version.bzl`.

File: tests/core/core.spec.ts (hunk 1 above)
Comment (a2wd) [reply tests/core/core.spec.ts:894]: Thanks, done

File: core/session.ts (hunk 2 above)
Comment (a2wd) [reply core/session.ts:444]: Sounds good, done - thanks.

Comment (a2wd) [reply core/session.ts:444]: Ack. Have updated alongside the `normalize` => `finalize` change.

**Comment by a2wd** [comment]
> Nice, looks great to me! But wait for ben to check.
> 
> Probably worth doing a patch version bump https://github.com/dataform-co/dataform/blob/main/version.bzl
> 
> Nit: do the refactoring part in a separate CL next time

Wasn't sure on the way we prefer to do this (eg. this is all in separate commits, but one PR). Will keep to separate PRs in the future, especially considering we squash on merge.

File: tests/core/core.spec.ts (hunk 3 above)
Comment (a2wd) [reply tests/core/core.spec.ts]: Thanks, updated

File: tests/core/core.spec.ts [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 import * as compilers from "df/core/compilers";
-import { Session } from "df/core/session";
-import { targetAsReadableString } from "df/core/targets";
-import { dataform } from "df/protos/ts";
-import { suite, test } from "df/testing";
-import { asPlainObject } from "df/tests/utils";
+import {Session} from "df/core/session";
+import {targetAsReadableString} from "df/core/targets";
+import {dataform} from "df/protos/ts";
+import {suite, Test, test} from "df/testing";
```
Comment (BenBirt) [inline tests/core/core.spec.ts:9]: do you need this `Test` import?

File: tests/core/core.spec.ts [hunk 5]
```diff
@@ -931,6 +946,17 @@ suite("@dataform/core", () => {
       ).deep.equals(["schema.operate-1"]);
       expect(graph.operations[1].queries).deep.equals(['select * from "schema"."operate-1"']);
     });
+
+    [{testConfig: TestConfigs.redshift, finalizedSchema: 'schema'}, {testConfig: TestConfigs.redshiftWithSuffix, finalizedSchema: 'schema_suffix'}].forEach(({testConfig, finalizedSchema}) => {
```
Comment (BenBirt) [inline tests/core/core.spec.ts:950]: seems bigger than line limit. ideally, prettier should be auto-formatting your code, can you look into that?

**Review by BenBirt — APPROVED** [review]

File: tests/core/core.spec.ts (hunk 5 above)
Comment (a2wd) [reply tests/core/core.spec.ts:950]: Did speak with Elias about that, he said we use the auto format in VS code, but that didn't touch line length. I did notice we have other tests with an `array.forEach` that goes past the standard 100 character width, so left that.

I suggest we add some global format file/tool (eg. a `tslint`) and update the contributing.md too, to mention using this. WDYT?

Comment (BenBirt) [reply tests/core/core.spec.ts:950]: We do use tslint in closed-source, I'm a little surprised it's not enforced here (as part of presubmit tests).

File: tests/core/core.spec.ts (hunk 4 above)
Comment (a2wd) [reply tests/core/core.spec.ts]: No, removed.

**Review by BenBirt — APPROVED** [review]

File: tests/core/core.spec.ts (hunk 5 above)
Comment (a2wd) [reply tests/core/core.spec.ts:950]: Will look into adding this.

**Pull request merged**